client/client-multi/client-c/src/kaa_logging.c

```diff
@@ -66,20 +71,95 @@ static const kaa_service_t logging_sync_services[1] = {KAA_SERVICE_LOGGING};
 
 
 
+static kaa_error_t remember_request(kaa_log_collector_t *self, uint16_t bucket_id)
+{
+    KAA_RETURN_IF_NIL(self, KAA_ERR_BADPARAM);
+
+    timeout_info_t *info = (timeout_info_t *)KAA_MALLOC(sizeof(timeout_info_t));
+    KAA_RETURN_IF_NIL(info, KAA_ERR_NOMEM);
+
+    info->log_bucket_id = bucket_id;
+    info->timeout = time(NULL) + (time_t)ext_log_upload_strategy_get_timeout(self->log_upload_strategy_context);
+
+    kaa_list_t *it = self->timeouts ? kaa_list_push_front(self->timeouts, info) : kaa_list_create(info);
+    KAA_RETURN_IF_NIL(it, KAA_ERR_NOMEM);
+
+    self->timeouts = it;
+
+    return KAA_ERR_NONE;
+}
+
+
+
+static bool find_by_bucket_id(void *data, void *context)
+{
+    KAA_RETURN_IF_NIL2(data, context, KAA_ERR_BADPARAM);
+    return (((timeout_info_t *)data)->log_bucket_id = *((uint16_t *)context));
+}
+
+
+
+static kaa_error_t remove_request(kaa_log_collector_t *self, uint16_t bucket_id)
+{
+    KAA_RETURN_IF_NIL(self, KAA_ERR_BADPARAM);
+    kaa_list_remove_first(&self->timeouts, &find_by_bucket_id, &bucket_id, NULL);
+    return KAA_ERR_NONE;
+}
+
+
+
+static bool is_timeout(kaa_log_collector_t *self)
+{
+    KAA_RETURN_IF_NIL2(self, self->timeouts, false);
+
+    bool is_timeout = false;
+    timeout_info_t *info;
+    kaa_list_t *it = self->timeouts;
+    time_t now = time(NULL);
```

Comment:
time - need to check Econais compatibility, may be we need to have platform specific time utilities.